Import resiliency (3/3)

BenEmdon opened the pull request:
## What
This PR creates a new user flow that utilizes the new import resiliency changes made in #1368 and #1378 for the purpose of solving #1253. 

This particular PR proposes a couple of significant additions:
* Registers `CreateGitHubRepositoryJob`'s and `PorterStatusJob`'s queues with sidekiq
* Configures ActionCable with our Redis setup
* Added a `status` column to `AssignmentInvitation`
* Introduces a new `import_resiliency` feature 🚩 
* The following routes on `AssignmentInvitationController`:
  * `GET #progress` - JSON api that returns the status of an invitation
  * `POST #create_repo` - if the invitation is `accepted`, kicks off a `CreateGitHubRepositoryJob`
  * `GET #setupv2` - renders a live updating view of the repo creation progress

All of the new repo creation flow is hiding behind the `import_resiliency` feature 🚩 

## ActionCable
All the server work for ActionCable is done. The client side work will require some extra thought so I have left that work for a follow up PR since this is already an 🐘. Instead of ActionCable we are using short polling and hitting a `GET #progress` for now.

## What if it errors?
Right now if it errors the UI will reflect the error state but I am yet to add retry button. There is a route ready for the retry button but I have a question about unique/locking background jobs.

## What does it look like?
![jul-10-2018 11-18-07](https://user-images.githubusercontent.com/11095731/42529579-b5b51bc8-8433-11e8-8955-897110d3e1a8.gif)
![jul-10-2018 11-18-41](https://user-images.githubusercontent.com/11095731/42529589-bc89e2da-8433-11e8-9d08-57d338ae96b5.gif)

BenEmdon commented on app/controllers/assignment_invitations_controller.rb:
```diff
@@ -8,19 +9,79 @@ class AssignmentInvitationsController < ApplicationController
+    if import_resiliency_enabled?
+      job_started = false
+      if current_invitation.accepted? || current_invitation.errored?
+        AssignmentRepo::CreateGitHubRepositoryJob.perform_later(current_assignment, current_user)
```
Should we ensure that the job is a unique job before we kick it off? Since jobs retry with exponential backoff we could end up in a situation where another job is kicked off while another is about to be retired.

BenEmdon replied:
By default Sidekiq doesn't have this feature. It does in enterprise, but we aren't on that plan.

Another option would be to use [mhenrixon/sidekiq-unique-jobs](https://github.com/mhenrixon/sidekiq-unique-jobs). This would require us to move away from ActiveJob and to using straight Sidekiq.

Possibly worth investigating in a follow up PR.

BenEmdon commented on app/jobs/assignment_repo/create_github_repository_job.rb:
```diff
@@ -56,12 +49,29 @@ def perform(assignment, user)
+          RepositoryCreationStatusChannel.channel(user_id: user.id),
+          text: Creator::REPOSITORY_CREATION_COMPLETE,
+          status: assignment.invitation&.status
+        )
```
Should we be recording stats about successful repo creation on this vs the old way? if so using  `GitHubClassroom.statsd`?

BenEmdon commented on config/sidekiq.yml:
```diff
@@ -5,10 +5,12 @@
+  - [deadline,          3]
+  - [boom,              3]
+  - [create_repository, 3]
+  - [porter_status,     3]
```
Worth mentioning that this many custom queues could be inefficient (#1381 )

d12 commented:
Also one other thing I'd suggest, this PR is pretty big and it's hard to review and be sure it's 100% correct. If it's possible in the future, I'd recommend breaking PRs into more atomic, smaller changes 😅

BenEmdon commented on app/controllers/assignment_invitations_controller.rb:
```diff
@@ -8,19 +9,79 @@ class AssignmentInvitationsController < ApplicationController
+  end
+
+  # rubocop:disable MethodLength
+  def create
```
Good point. `create` doesn't make sense. What are your thoughts on the actions instead following:
```ruby
patch :create_assignment
patch :create_repo
patch :create_assignment_repo
```

BenEmdon commented on app/models/assignment_invitation.rb:
```diff
@@ -20,12 +22,18 @@ class AssignmentInvitation < ApplicationRecord
+  #
+  # rubocop:disable MethodLength
+  # rubocop:disable AbcSize
+  def redeem_for(invitee, import_resiliency: false)
```
I can't use the feature dependency here because we are outside of a controller. (feature 🚩s are defined on `ApplicationController`).

BenEmdon commented:
## Update
* Changed `#create` -> `#create_repo`
* Moved poll interval out into a constant
* Merged ActionCable fix in #1389 
* Merged migration in #1386 

## Things to do in followup PRs
* Fix concurrency issue with `CreateGitHubRepositoryJob`
* Add support for group assignments
* Support ActionCable instead of short polling

Can I get some final 👀 on this? @d12
